File: android_app/app/src/main/java/com/sikwin/app/utils/UnityTokenHelper.java
package com.sikwin.app.utils;

import android.util.Log;
import org.json.JSONObject;

public class UnityTokenHelper {
    
    /**
     * Send authentication tokens to Unity GameManager using reflection
     * This should be called after successful login in Kotlin
     */
    public static void sendTokensToUnity(String access, String refresh, String username) {
        try {
            JSONObject json = new JSONObject();
            json.put("access", access);
            json.put("refresh", refresh);
            json.put("username", username != null ? username : "");
            final String jsonString = json.toString();
            
            Log.d("UnityTokenHelper", "Sending tokens to Unity: GameManager");
            
            // Run on UI thread to ensure UnityPlayer is accessible
            android.os.Handler handler = new android.os.Handler(android.os.Looper.getMainLooper());
            handler.post(new Runnable() {
                @Override
                public void run() {
                    try {
                        // Use reflection to access UnityPlayer since it's in a different module
                        Class<?> unityPlayerClass = Class.forName("com.unity3d.player.UnityPlayer");
                        java.lang.reflect.Method unitySendMessage = unityPlayerClass.getMethod(
                            "UnitySendMessage", 
                            String.class, 
                            String.class, 
                            String.class
                        );
                        
                        // Send to GameManager (primary target)
                        unitySendMessage.invoke(null, "GameManager", "SetAccessAndRefreshTokens", jsonString);
                        unitySendMessage.invoke(null, "GameManager", "SetToken", access);
                        
                        Log.d("UnityTokenHelper", "Tokens sent successfully to Unity via reflection");
                    } catch (ClassNotFoundException e) {
                        Log.d("UnityTokenHelper", "UnityPlayer not available yet");
                    } catch (Exception e) {
                        Log.e("UnityTokenHelper", "Error in UI thread sending tokens: " + e.getMessage());
                    }
                }
            });
        } catch (Exception e) {
            Log.e("UnityTokenHelper", "Error preparing tokens for Unity: " + e.getMessage(), e);
        }
    }

    /**
     * Trigger logout in Unity GameManager
     */
    public static void sendLogoutToUnity() {
        try {
            Log.d("UnityTokenHelper", "Sending logout to Unity: GameManager");
            
            android.os.Handler handler = new android.os.Handler(android.os.Looper.getMainLooper());
            handler.post(new Runnable() {
                @Override
                public void run() {
                    try {
                        Class<?> unityPlayerClass = Class.forName("com.unity3d.player.UnityPlayer");
                        java.lang.reflect.Method unitySendMessage = unityPlayerClass.getMethod(
                            "UnitySendMessage", 
                            String.class, 
                            String.class, 
                            String.class
                        );
                        
                        unitySendMessage.invoke(null, "GameManager", "Logout", "");
                        Log.d("UnityTokenHelper", "Logout signal sent successfully to Unity");
                    } catch (Exception e) {
                        Log.e("UnityTokenHelper", "Error sending logout to Unity: " + e.getMessage());
                    }
                }
            });
        } catch (Exception e) {
            Log.e("UnityTokenHelper", "Error preparing logout for Unity: " + e.getMessage());
        }
    }
}

File: android_app/app/src/main/java/com/sikwin/app/utils/UnityTokenHelper.kt
package com.sikwin.app.utils

import android.content.Context
import android.content.Intent
import android.util.Log
import org.json.JSONObject

object UnityTokenHelper {
    private const val TAG = "UnityTokenHelper"

    /**
     * Send access and refresh tokens to Unity.
     */
    fun sendTokensToUnity(access: String, refresh: String) {
        // Disabled access token passing as requested
        Log.d(TAG, "sendTokensToUnity: Access token passing is currently disabled")
        return
    }

    /**
     * Sends tokens to Unity: direct UnitySendMessage + broadcast fallback.
     * Token-only: do NOT send username/password.
     */
    fun sendTokensToUnity(
        context: Context,
        access: String,
        refresh: String
    ) {
        try {
            // Avoid UnitySendMessage token injection; rely on prefs pre-write and broadcast.
            sendTokensToUnity(access, refresh)

            val intent = Intent("com.sikwin.app.TOKEN_UPDATE").apply {
                putExtra("access", access)
                putExtra("refresh", refresh)
                setPackage(context.packageName)
            }
            context.sendBroadcast(intent)
            Log.d(TAG, "Token broadcast sent")
        } catch (e: Exception) {
            Log.e(TAG, "Error sending token broadcast: ${e.message}")
        }
    }

    /**
     * Trigger logout in Unity using Broadcast.
     */
    fun sendLogoutToUnity(context: Context) {
        try {
            Log.d(TAG, "Sending logout to Unity via Broadcast")
            val intent = Intent("com.sikwin.app.TOKEN_UPDATE").apply {
                putExtra("action", "logout")
                setPackage(context.packageName)
            }
            context.sendBroadcast(intent)
        } catch (e: Exception) {
            Log.e(TAG, "Error sending logout broadcast: ${e.message}")
        }
    }
}
